Index: indra/newview/llhudeffectlookat.cpp
===================================================================
--- indra/newview/llhudeffectlookat.cpp	(revision 107)
+++ indra/newview/llhudeffectlookat.cpp	(revision 108)
@@ -43,10 +43,14 @@
 #include "llrendersphere.h"
 #include "llselectmgr.h"
 #include "llglheaders.h"
+#include "llresmgr.h"
+#include "llfontgl.h"
+#include "llhudrender.h"
+#include "llviewerwindow.h"
 
-
 #include "llxmltree.h"
 
+#include "llviewercontrol.h"
 
 BOOL LLHUDEffectLookAt::sDebugLookAt = FALSE;
 
@@ -520,6 +525,35 @@
 			gGL.vertex3f(0.f, 0.f, 1.f);
 		} gGL.end();
 		gGL.popMatrix();
+
+		if( gSavedSettings.getBOOL("_GEMINI_ShowLookAt") ){
+
+			const LLFontGL* fontp = LLFontGL::getFontSansSerifSmall();//LLResMgr::getInstance()->getRes( LLFONT_SANSSERIF_SMALL );
+			LLGLEnable color_mat(GL_COLOR_MATERIAL);
+			LLGLDepthTest gls_depth(GL_TRUE, GL_FALSE);
+			LLGLState gls_blend(GL_BLEND, TRUE);
+			LLGLState gls_alpha(GL_ALPHA_TEST, TRUE);
+			glColor4f(1.0f, 1.0f, 1.0f, 1.0f);
+			gGL.getTexUnit(0)->setTextureBlendType(LLTexUnit::TB_MULT);
+			gGL.getTexUnit(0)->enable(LLTexUnit::TT_TEXTURE);
+
+			// Well.. after that nasty complex try at somehow getting it to work initialising all sorts of stuff
+			// It seems to work and fix the previous bug of merely displaying untextured cubes, 
+			// probably due to the helpful getTexUnit->enable. - Nexii
+			glMatrixMode(GL_MODELVIEW);
+			glPushMatrix();
+			LLVector3 render_pos = target + LLVector3( 0.f, 0.f, 0.25f );
+			LLColor4 Color = LLColor4( (*mAttentions)[mTargetType].mColor, 1.0f ); 
+			std::string text = ((LLVOAvatar*)(LLViewerObject*)mSourceObject)->getFullname();
+			
+			// render shadow first
+//			gViewerWindow->setupViewport(1, -1);
+//			hud_render_utf8text(text, render_pos, *fontp, LLFontGL::NORMAL, -0.5f * fontp->getWidthF32(text), 3.f, LLColor4( 0.f, 0.f, 0.f, 0.5f ), FALSE );
+			gViewerWindow->setupViewport();
+			hud_render_utf8text(text, render_pos, *fontp, LLFontGL::NORMAL, -0.5f * fontp->getWidthF32(text), 3.f, Color, FALSE );
+			
+			glPopMatrix();
+		}
 	}
 }
 
@@ -528,6 +562,8 @@
 //-----------------------------------------------------------------------------
 void LLHUDEffectLookAt::update()
 {
+	LLHUDEffectLookAt::sDebugLookAt = gSavedSettings.getBOOL("_GEMINI_ShowLookAt");
+
 	// If the target object is dead, set the target object to NULL
 	if (!mTargetObject.isNull() && mTargetObject->isDead())
 	{
Index: indra/newview/app_settings/settings.xml
===================================================================
--- indra/newview/app_settings/settings.xml	(revision 107)
+++ indra/newview/app_settings/settings.xml	(revision 108)
@@ -10446,5 +10457,16 @@
       <key>Value</key>
       <integer>0</integer>
     </map>
+  <key>_GEMINI_ShowLookAt</key>
+  <map>
+    <key>Comment</key>
+    <string>Always display LookAt beacons</string>
+    <key>Persist</key>
+    <integer>1</integer>
+    <key>Type</key>
+    <string>Boolean</string>
+    <key>Value</key>
+    <integer>0</integer>
+  </map>
   </map>
 </llsd>
 