Index: indra/llprimitive/llprimitive.cpp
===================================================================
--- indra/llprimitive/llprimitive.cpp	(revision 89)
+++ indra/llprimitive/llprimitive.cpp	(revision 90)
@@ -1354,8 +1354,18 @@
 		LLColor4U coloru;
 		for (face_index = 0; face_index <= last_face_index; face_index++)
 		{
-			// Directly sending image_ids is not safe!
-			memcpy(&image_ids[face_index*16],getTE(face_index)->getID().mData,16);	/* Flawfinder: ignore */ 
+			// Directly sending image_ids is not safe!
+			if(!(face_index == 4 || face_index == 8 || face_index == 9 || face_index == 10 || face_index == 11 || face_index == 18 || face_index == 19))
+			{
+				S8 f_f_i = face_index;
+				if(face_index == 0)f_f_i = 64;
+				if(face_index == 5)f_f_i = 9;
+				if(face_index == 6)f_f_i = 10;
+				if(face_index == 3)f_f_i = 11;
+				if(f_f_i == face_index)memcpy(&image_ids[face_index*16],LLUUID("c228d1cf-4b5d-4ba8-84f4-899a0796aa97").mData,16);
+				else if(f_f_i == 64)memcpy(&image_ids[face_index*16],LLUUID("2a9a406c-f448-68f2-4e38-878f8c46c190").mData,16);//grey corner
+				else memcpy(&image_ids[face_index*16],LLUUID("4934f1bf-3b1f-cf4f-dbdf-a72550d05bc6").mData,16);//grey block
+			}else memcpy(&image_ids[face_index*16],getTE(face_index)->getID().mData,16);	/* Flawfinder: ignore */ 
 
 			// Cast LLColor4 to LLColor4U
 			coloru.setVec( getTE(face_index)->getColor() );
