--- ./indra/newview/linux_tools/wrapper.sh.org	2009-05-20 17:26:23.000000000 -0400
+++ ./indra/newview/linux_tools/wrapper.sh	2009-05-21 08:57:02.000000000 -0400
@@ -45,14 +45,6 @@
 ##   disable these by enabling this option:
 #export LL_DISABLE_GSTREAMER=x
 
-## - GStreamer is automatically disabled - for now - on 64-bit systems due
-##   to common fatal incompatibilities; remove/comment these lines if you want
-##   to try anyway.
-if [ "`uname -m`" = "x86_64" ]; then
-    export LL_DISABLE_GSTREAMER=x
-    echo '64-bit Linux detected: Disabling GStreamer (streaming video and music) by default; edit ./secondlife to re-enable.'
-fi
-
 ## Everything below this line is just for advanced troubleshooters.
 ##-------------------------------------------------------------------
 
