Index: trunk-clean/indra/newview/skins/default/xui/en-us/floater_joystick.xml
===================================================================
--- trunk-clean/indra/newview/skins/default/xui/en-us/floater_joystick.xml	(revision 1135)
+++ trunk-clean/indra/newview/skins/default/xui/en-us/floater_joystick.xml	(working copy)
@@ -110,10 +110,17 @@
 	<text bottom="-450" left="20" width="94" halign="right" name="ZoomDeadZone">Zoom Dead Zone</text>
 	<spinner bottom="-450" left="265" width="56" label_width="0" control_name="FlycamAxisDeadZone6" name="FlycamAxisDeadZone6" label="" decimal_digits="2" increment="0.01" min_val="0" max_val="1"/>
 
-	<button bottom="-451" left="336" enabled="true" follows="left|top" font="SansSerif"
+	<button bottom="-451" left="340" enabled="true" follows="left|top" font="SansSerif"
 	     halign="center" height="22" hidden="false"
 	     label="SpaceNavigator Defaults" mouse_opaque="true"
-	     name="SpaceNavigatorDefaults" scale_image="true" width="200" />
+	     name="SpaceNavigatorDefaults" scale_image="true" width="184" />
+	
+	<button bottom="-480" follows="right|bottom" font="SansSerif" halign="center"
+	     height="20" label="OK" label_selected="Cancel" left="340"
+	     mouse_opaque="true" name="ok_btn" width="90" />
+	<button bottom="-480" follows="right|bottom" font="SansSerif" halign="center"
+	     height="20" label="Cancel" label_selected="Select" left_delta="94"
+	     mouse_opaque="true" name="cancel_btn" width="90" />
 
 	<string name="JoystickMonitor">Joystick Monitor</string>
 	<string name="Axis">Axis [NUM]</string>
Index: trunk-clean/indra/newview/llfloaterjoystick.h
===================================================================
--- trunk-clean/indra/newview/llfloaterjoystick.h	(revision 1135)
+++ trunk-clean/indra/newview/llfloaterjoystick.h	(working copy)
@@ -50,15 +50,21 @@
 
 private:
 	static void onClickRestoreSNDefaults(void*);
+	static void onClickCancel(void*);
+	static void onClickOK(void*);
 
 private:
 	// Device prefs
+	bool mJoystickEnabled;
 	S32 mJoystickAxis[7];
 	bool m3DCursor;
 	bool mAutoLeveling;
 	bool mZoomDirect;
 
 	// Modes prefs
+	bool mAvatarEnabled;
+	bool mBuildEnabled;
+	bool mFlycamEnabled;
 	F32 mAvatarAxisScale[6];
 	F32 mBuildAxisScale[6];
 	F32 mFlycamAxisScale[7];
Index: trunk-clean/indra/newview/llfloaterjoystick.cpp
===================================================================
--- trunk-clean/indra/newview/llfloaterjoystick.cpp	(revision 1135)
+++ trunk-clean/indra/newview/llfloaterjoystick.cpp	(working copy)
@@ -115,6 +115,9 @@
 	addChild(mAxisStatsView);
 	
 	childSetAction("SpaceNavigatorDefaults", onClickRestoreSNDefaults, this);
+	childSetAction("cancel_btn", onClickCancel, this);
+	childSetAction("ok_btn", onClickOK, this);
+
 	refresh();
 	return TRUE;
 }
@@ -133,6 +136,8 @@
 {
 	LLFloater::refresh();
 
+	mJoystickEnabled = gSavedSettings.getBOOL("JoystickEnabled");
+
 	mJoystickAxis[0] = gSavedSettings.getS32("JoystickAxis0");
 	mJoystickAxis[1] = gSavedSettings.getS32("JoystickAxis1");
 	mJoystickAxis[2] = gSavedSettings.getS32("JoystickAxis2");
@@ -145,6 +148,10 @@
 	mAutoLeveling = gSavedSettings.getBOOL("AutoLeveling");
 	mZoomDirect  = gSavedSettings.getBOOL("ZoomDirect");
 
+	mAvatarEnabled = gSavedSettings.getBOOL("JoystickAvatarEnabled");
+	mBuildEnabled = gSavedSettings.getBOOL("JoystickBuildEnabled");
+	mFlycamEnabled = gSavedSettings.getBOOL("JoystickFlycamEnabled");
+	
 	mAvatarAxisScale[0] = gSavedSettings.getF32("AvatarAxisScale0");
 	mAvatarAxisScale[1] = gSavedSettings.getF32("AvatarAxisScale1");
 	mAvatarAxisScale[2] = gSavedSettings.getF32("AvatarAxisScale2");
@@ -196,9 +205,7 @@
 
 void LLFloaterJoystick::cancel()
 {
-	llinfos << "reading from gSavedSettings->Cursor3D=" 
-		<< gSavedSettings.getBOOL("Cursor3D") << "; m3DCursor=" 
-		<< m3DCursor << llendl;
+	gSavedSettings.setBOOL("JoystickEnabled", mJoystickEnabled);
 
 	gSavedSettings.setS32("JoystickAxis0", mJoystickAxis[0]);
 	gSavedSettings.setS32("JoystickAxis1", mJoystickAxis[1]);
@@ -212,6 +219,10 @@
 	gSavedSettings.setBOOL("AutoLeveling", mAutoLeveling);
 	gSavedSettings.setBOOL("ZoomDirect", mZoomDirect );
 
+	gSavedSettings.setBOOL("JoystickAvatarEnabled", mAvatarEnabled);
+	gSavedSettings.setBOOL("JoystickBuildEnabled", mBuildEnabled);
+	gSavedSettings.setBOOL("JoystickFlycamEnabled", mFlycamEnabled);
+	
 	gSavedSettings.setF32("AvatarAxisScale0", mAvatarAxisScale[0]);
 	gSavedSettings.setF32("AvatarAxisScale1", mAvatarAxisScale[1]);
 	gSavedSettings.setF32("AvatarAxisScale2", mAvatarAxisScale[2]);
@@ -266,6 +277,33 @@
 	setSNDefaults();
 }
 
+void LLFloaterJoystick::onClickCancel(void *joy_panel)
+{
+	if (joy_panel)
+	{
+		LLFloaterJoystick* self = (LLFloaterJoystick*)joy_panel;
+
+		if (self)
+		{
+			self->cancel();
+			self->close();
+		}
+	}
+}
+
+void LLFloaterJoystick::onClickOK(void *joy_panel)
+{
+	if (joy_panel)
+	{
+		LLFloaterJoystick* self = (LLFloaterJoystick*)joy_panel;
+
+		if (self)
+		{
+			self->close();
+		}
+	}
+}
+
 void LLFloaterJoystick::setSNDefaults()
 {
 	LLViewerJoystick::getInstance()->setSNDefaults();
Index: trunk-clean/doc/contributions.txt
===================================================================