Index: newview/CMakeLists.txt
===================================================================
--- newview/CMakeLists.txt      (revision 109755)
+++ newview/CMakeLists.txt      (working copy)
@@ -1418,6 +1418,7 @@
     ${LSCRIPT_LIBRARIES}
     ${LLMATH_LIBRARIES}
     ${LLCOMMON_LIBRARIES}
+    ${NDOF_LIBRARY}
     ${viewer_LIBRARIES}
     ${BOOST_PROGRAM_OPTIONS_LIBRARY}
     ${BOOST_REGEX_LIBRARY}
@@ -1429,7 +1430,6 @@
     ${SDL_LIBRARY}
     ${SMARTHEAP_LIBRARY}
     ${UI_LIBRARIES}
-    ${NDOF_LIBRARY}
     ${QUICKTIME_LIBRARY}
     ${WINDOWS_LIBRARIES}
     ${XMLRPCEPI_LIBRARIES}