diff -urN linden/indra/newview/lltexturecache.cpp linden-patched/indra/newview/lltexturecache.cpp
--- linden/indra/newview/lltexturecache.cpp	2009-03-11 18:46:04.000000000 +0100
+++ linden-patched/indra/newview/lltexturecache.cpp	2009-04-15 10:28:03.000000000 +0200
@@ -1471,19 +1471,14 @@
 		purgeTextures(false);
 		mDoPurge = FALSE;
 	}
-	if (datasize >= TEXTURE_CACHE_ENTRY_SIZE)
-	{
-		LLMutexLock lock(&mWorkersMutex);
-		llassert_always(imagesize > 0);
-		LLTextureCacheWorker* worker = new LLTextureCacheRemoteWorker(this, priority, id,
+	LLMutexLock lock(&mWorkersMutex);
+	llassert_always(imagesize > 0);
+	LLTextureCacheWorker* worker = new LLTextureCacheRemoteWorker(this, priority, id,
 																data, datasize, 0,
 																imagesize, responder);
-		handle_t handle = worker->write();
-		mWriters[handle] = worker;
-		return handle;
-	}
-	delete responder;
-	return LLWorkerThread::nullHandle();
+	handle_t handle = worker->write();
+	mWriters[handle] = worker;
+	return handle;
 }
 
 bool LLTextureCache::writeComplete(handle_t handle, bool abort)
