diff -urN linden/indra/newview/llappviewer.cpp linden-patched/indra/newview/llappviewer.cpp
--- linden/indra/newview/llappviewer.cpp	2008-10-17 17:55:32.000000000 +0200
+++ linden-patched/indra/newview/llappviewer.cpp	2008-10-18 10:44:12.000000000 +0200
@@ -3648,7 +3648,7 @@
 	// Check that the circuit between the viewer and the agent's current
 	// region is still alive
 	LLViewerRegion *agent_region = gAgent.getRegion();
-	if (agent_region)
+	if (agent_region && LLStartUp::getStartupState() == STATE_STARTED)
 	{
 		LLUUID this_region_id = agent_region->getRegionID();
 		bool this_region_alive = agent_region->isAlive();
