diff -urN linden/indra/llcharacter/llkeyframewalkmotion.cpp linden-patched/indra/llcharacter/llkeyframewalkmotion.cpp
--- linden/indra/llcharacter/llkeyframewalkmotion.cpp	2007-11-08 03:18:06.000000000 +0100
+++ linden-patched/indra/llcharacter/llkeyframewalkmotion.cpp	2007-11-15 23:21:50.000000000 +0100
@@ -49,6 +49,7 @@
 const F32 MAX_TIME_DELTA = 2.f;		//max two seconds a frame for calculating interpolation
 const F32 SPEED_ADJUST_MAX = 2.5f; // maximum adjustment of walk animation playback speed
 const F32 SPEED_ADJUST_MAX_SEC = 3.f;	// maximum adjustment to walk animation playback speed for a second
+const F32 SPEED_FINAL_SCALING = 0.5f;	// final scaling for walk animation
 const F32 DRIFT_COMP_MAX_TOTAL = 0.07f;//0.55f; // maximum drift compensation overall, in any direction 
 const F32 DRIFT_COMP_MAX_SPEED = 4.f; // speed at which drift compensation total maxes out
 const F32 MAX_ROLL = 0.6f;
@@ -304,6 +305,7 @@
 	}
 
 	mAnimSpeed = (mAvgSpeed + mSpeedAdjust) * mRelativeDir;
+	mAnimSpeed = mAnimSpeed * SPEED_FINAL_SCALING;
 //	char debug_text[64];
 //	sprintf(debug_text, "Foot slip vel: %.2f", footSlipVelocity);
 //	mCharacter->addDebugText(debug_text);
