diff -urN linden/indra/newview/llinventorybridge.cpp linden-patched/indra/newview/llinventorybridge.cpp
--- linden/indra/newview/llinventorybridge.cpp	2008-02-29 19:47:45.000000000 +0100
+++ linden-patched/indra/newview/llinventorybridge.cpp	2009-02-11 11:50:41.000000000 +0100
@@ -3235,10 +3235,19 @@
 
 void LLObjectBridge::openItem()
 {
-	/* Disabled -- this preview isn't useful. JC */
-	// CP: actually, this code is required - made changes to match LLAnimationBridge::openItem() idiom
-	// The properties preview is useful, converting to show object properties. - DaveP
-	LLShowProps::showProperties(mUUID);
+	LLVOAvatar* avatar = gAgent.getAvatarObject();
+	if (!avatar)
+	{
+		return;
+	}
+	if (avatar->isWearingAttachment(mUUID))
+	{
+		performAction(NULL, NULL, "detach");
+	}
+	else
+	{
+		performAction(NULL, NULL, "attach");
+	}
 }
 
 LLFontGL::StyleFlags LLObjectBridge::getLabelStyle() const
@@ -4268,9 +4277,13 @@
 	}
 	else if(isAgentInventory())
 	{
-		if( !gAgent.isWearingItem( mUUID ) )
+		if (gAgent.isWearingItem(mUUID))
 		{
-			wearOnAvatar();
+			performAction(NULL, NULL, "take_off");
+		}
+		else
+ 		{
+			performAction(NULL, NULL, "wear");
 		}
 	}
 	else
