diff -urN linden/indra/llmessage/llhttpclient.cpp linden-patched/indra/llmessage/llhttpclient.cpp
--- linden/indra/llmessage/llhttpclient.cpp	2008-10-17 17:55:31.000000000 +0200
+++ linden-patched/indra/llmessage/llhttpclient.cpp	2008-10-18 12:20:45.000000000 +0200
@@ -160,10 +160,9 @@
 				fstream.seekg(0, std::ios::end);
 				U32 fileSize = fstream.tellg();
 				fstream.seekg(0, std::ios::beg);
-				char* fileBuffer;
-				fileBuffer = new char [fileSize];
-				fstream.read(fileBuffer, fileSize);
-				ostream.write(fileBuffer, fileSize);
+				std::vector<char> fileBuffer(fileSize);
+				fstream.read(&fileBuffer[0], fileSize);
+				ostream.write(&fileBuffer[0], fileSize);
 				fstream.close();
 				eos = true;
 				return STATUS_DONE;
@@ -190,10 +189,9 @@
 			
 			LLVFile vfile(gVFS, mUUID, mAssetType, LLVFile::READ);
 			S32 fileSize = vfile.getSize();
-			U8* fileBuffer;
-			fileBuffer = new U8 [fileSize];
-            vfile.read(fileBuffer, fileSize);
-            ostream.write((char*)fileBuffer, fileSize);
+			std::vector<U8> fileBuffer(fileSize);
+			vfile.read(&fileBuffer[0], fileSize);
+			ostream.write((char*)&fileBuffer[0], fileSize);
 			eos = true;
 			return STATUS_DONE;
 		}
