diff -urN linden/indra/newview/llfloaterland.cpp linden-patched/indra/newview/llfloaterland.cpp
--- linden/indra/newview/llfloaterland.cpp	2008-02-29 19:47:45.000000000 +0100
+++ linden-patched/indra/newview/llfloaterland.cpp	2008-10-18 11:25:46.000000000 +0200
@@ -106,7 +106,7 @@
 // LLFloaterLand
 //---------------------------------------------------------------------------
 
-void send_parcel_select_objects(S32 parcel_local_id, S32 return_type,
+void send_parcel_select_objects(S32 parcel_local_id, U32 return_type,
 								uuid_list_t* return_ids = NULL)
 {
 	LLMessageSystem *msg = gMessageSystem;
@@ -124,7 +124,7 @@
 	msg->addUUIDFast(_PREHASH_SessionID,gAgent.getSessionID());
 	msg->nextBlockFast(_PREHASH_ParcelData);
 	msg->addS32Fast(_PREHASH_LocalID, parcel_local_id);
-	msg->addS32Fast(_PREHASH_ReturnType, return_type);
+	msg->addU32Fast(_PREHASH_ReturnType, return_type);
 
 	// Throw all return ids into the packet.
 	// TODO: Check for too many ids.
diff -urN linden/indra/newview/llfloatersellland.cpp linden-patched/indra/newview/llfloatersellland.cpp
--- linden/indra/newview/llfloatersellland.cpp	2008-02-29 19:47:45.000000000 +0100
+++ linden-patched/indra/newview/llfloatersellland.cpp	2008-10-18 11:26:10.000000000 +0200
@@ -46,7 +46,7 @@
 #include "llviewerwindow.h"
 
 // defined in llfloaterland.cpp
-void send_parcel_select_objects(S32 parcel_local_id, S32 return_type,
+void send_parcel_select_objects(S32 parcel_local_id, U32 return_type,
 								uuid_list_t* return_ids = NULL);
 
 enum Badge { BADGE_OK, BADGE_NOTE, BADGE_WARN, BADGE_ERROR };
