diff -urN linden/indra/newview/app_settings/keywords.ini linden-patched/indra/newview/app_settings/keywords.ini
--- linden/indra/newview/app_settings/keywords.ini	2008-10-17 17:55:32.000000000 +0200
+++ linden-patched/indra/newview/app_settings/keywords.ini	2008-10-18 10:47:46.000000000 +0200
@@ -455,6 +455,7 @@
 PARCEL_FLAG_ALLOW_TERRAFORM			Used with llGetParcelFlags to find if a parcel allows anyone to terraform the land
 PARCEL_FLAG_ALLOW_DAMAGE			Used with llGetParcelFlags to find if a parcel allows damage
 PARCEL_FLAG_ALLOW_CREATE_OBJECTS	Used with llGetParcelFlags to find if a parcel allows anyone to create objects
+PARCEL_FLAG_ALLOW_CREATE_GROUP_OBJECTS	Used with llGetParcelFlags to find if a parcel allows group members or objects to create objects
 PARCEL_FLAG_USE_ACCESS_GROUP		Used with llGetParcelFlags to find if a parcel limits access to a group
 PARCEL_FLAG_USE_ACCESS_LIST			Used with llGetParcelFlags to find if a parcel limits access to a list of residents
 PARCEL_FLAG_USE_BAN_LIST			Used with llGetParcelFlags to find if a parcel uses a ban list
