diff -r 827a910542a9 indra/llrender/llglslshader.cpp
--- a/indra/llrender/llglslshader.cpp	Mon Aug 17 14:54:41 2015 -0400
+++ b/indra/llrender/llglslshader.cpp	Sat Sep 19 10:44:14 2015 -0400
@@ -273,10 +273,13 @@
     glEndQueryARB(GL_TIME_ELAPSED);
     glEndQueryARB(GL_SAMPLES_PASSED);
     
-    U64 time_elapsed = 0;
+    //<FS:TS> U64 and GLuint64 somehow turn out different on x86_64
+    //U64 time_elapsed = 0;
+    GLuint64 time_elapsed = 0;
     glGetQueryObjectui64v(mTimerQuery, GL_QUERY_RESULT, &time_elapsed);
 
-    U64 samples_passed = 0;
+    //U64 samples_passed = 0;
+    GLuint64 samples_passed = 0;
     glGetQueryObjectui64v(mSamplesQuery, GL_QUERY_RESULT, &samples_passed);
 
     sTotalTimeElapsed += time_elapsed;
