diff -r 7b33b9057b22 indra/CMakeLists.txt
--- a/indra/CMakeLists.txt	Sat Feb 11 23:12:05 2012 +0100
+++ b/indra/CMakeLists.txt	Sun Feb 12 11:22:51 2012 -0500
@@ -69,8 +69,10 @@
   add_subdirectory(${LIBS_OPEN_PREFIX}llxuixml)
   add_subdirectory(${LIBS_OPEN_PREFIX}viewer_components)
 
-  # Legacy C++ tests. Build always, run if LL_TESTS is true.
-  add_subdirectory(${VIEWER_PREFIX}test)
+  if (LL_TESTS OR NOT STANDALONE)
+    # Legacy C++ tests. Build always, run if LL_TESTS is true.
+    add_subdirectory(${VIEWER_PREFIX}test)
+  endif (LL_TESTS OR NOT STANDALONE)
 
   # viewer media plugins
   add_subdirectory(${LIBS_OPEN_PREFIX}media_plugins)
